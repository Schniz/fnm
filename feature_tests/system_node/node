#!/bin/bash

echo "custom build"
