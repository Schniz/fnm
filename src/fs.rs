use std::path::Path;

#[cfg(unix)]
pub fn symlink_dir<P: AsRef<Path>, U: AsRef<Path>>(from: P, to: U) -> std::io::Result<()> {
    std::os::unix::fs::symlink(from, to)?;
    Ok(())
}

#[cfg(windows)]
pub fn symlink_dir<P: AsRef<Path>, U: AsRef<Path>>(from: P, to: U) -> std::io::Result<()> {
    junction::create(from, to)?;
    Ok(())
}

#[cfg(windows)]
pub fn remove_symlink_dir<P: AsRef<Path>>(path: P) -> std::io::Result<()> {
    std::fs::remove_dir(path)?;
    Ok(())
}

#[cfg(unix)]
pub fn remove_symlink_dir<P: AsRef<Path>>(path: P) -> std::io::Result<()> {
    std::fs::remove_file(path)?;
    Ok(())
}

pub fn shallow_read_symlink<P: AsRef<Path>>(path: P) -> std::io::Result<std::path::PathBuf> {
    std::fs::read_link(path)
}
