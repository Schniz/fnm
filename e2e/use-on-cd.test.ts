import { writeFile, mkdir } from "node:fs/promises"
import { join } from "node:path"
import { script } from "./shellcode/script.js"
import { Bash, Fish, PowerShell, WinCmd, Zsh } from "./shellcode/shells.js"
import testCwd from "./shellcode/test-cwd.js"
import testNodeVersion from "./shellcode/test-node-version.js"
import describe from "./describe.js"

for (const shell of [Bash, Zsh, Fish, PowerShell, WinCmd]) {
  describe(shell, () => {
    test(`use on cd`, async () => {
      await mkdir(join(testCwd(), "subdir"), { recursive: true })
      await writeFile(join(testCwd(), "subdir", ".node-version"), "v12.22.12")
      await script(shell)
        .then(shell.env({ useOnCd: true }))
        .then(shell.call("fnm", ["install", "v8.11.3"]))
        .then(shell.call("fnm", ["install", "v12.22.12"]))
        .then(shell.call("cd", ["subdir"]))
        .then(testNodeVersion(shell, "v12.22.12"))
        .takeSnapshot(shell)
        .execute(shell)
    })

    test(`with resolve-engines`, async () => {
      await mkdir(join(testCwd(), "subdir"), { recursive: true })
      await writeFile(
        join(testCwd(), "subdir", "package.json"),
        JSON.stringify({
          name: "hello",
          engines: {
            node: "v12.22.12",
          },
        }),
      )
      await script(shell)
        .then(shell.env({ useOnCd: true, resolveEngines: true }))
        .then(shell.call("fnm", ["install", "v8.11.3"]))
        .then(shell.call("fnm", ["install", "v12.22.12"]))
        .then(shell.call("cd", ["subdir"]))
        .then(testNodeVersion(shell, "v12.22.12"))
        .execute(shell)
    })
  })
}
