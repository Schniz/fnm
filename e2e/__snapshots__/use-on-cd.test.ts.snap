// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bash use on cd: Bash 1`] = `
"set -e
eval "$(fnm env --use-on-cd)"
fnm install v8.11.3
fnm install v12.22.12
cd subdir
if [ "$(node --version)" != "v12.22.12" ]; then
  echo "Expected node version to be v12.22.12. Got $(node --version)"
  exit 1
fi"
`;

exports[`Fish use on cd: Fish 1`] = `
"fnm env --use-on-cd | source
fnm install v8.11.3
fnm install v12.22.12
cd subdir
set ____test____ (node --version)
if test "$____test____" != "v12.22.12"
  echo "Expected node version to be v12.22.12. Got $____test____"
  exit 1
end"
`;

exports[`PowerShell use on cd: PowerShell 1`] = `
"$ErrorActionPreference = "Stop"
fnm env --use-on-cd | Out-String | Invoke-Expression
fnm install v8.11.3
fnm install v12.22.12
cd subdir
if ( "$(node --version)" -ne "v12.22.12" ) { echo "Expected node version to be v12.22.12. Got $(node --version)"; exit 1 }"
`;

exports[`Zsh use on cd: Zsh 1`] = `
"set -e
eval "$(fnm env --use-on-cd)"
fnm install v8.11.3
fnm install v12.22.12
cd subdir
if [ "$(node --version)" != "v12.22.12" ]; then
  echo "Expected node version to be v12.22.12. Got $(node --version)"
  exit 1
fi"
`;
